### PR TITLE
chore: booking verification token and booking rejection logic from email

### PR DESCRIPTION
**Pull request by anikdhabal**
## What does this PR do?

- Fixes #XXXX (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A-I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] N/A-I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: apps/web/pages/api/verify-booking-token.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+const querySchema = z.object({
+  action: z.nativeEnum(DirectAction),
+  token: z.string(),
+  reason: z.string().optional(),
+  bookingUid: z.string(),
+  userId: z.string(),
+});
+
+async function handler(req: NextApiRequest, res: NextApiResponse<Response>) {
+  const { action, token, reason, bookingUid, userId } = querySchema.parse(req.query);
```
Comment (zomars) [inline apps/web/pages/api/verify-booking-token.ts:27]: How are we getting the reason here? We cannot provide it from email AFAIK. Unless que put an [actual form in the email](https://www.mailmodo.com/guides/html-form-emails/) and transform this endpoint into a POST request.

File: apps/web/pages/api/verify-booking-token.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  ACCEPT = "accept",
+  REJECT = "reject",
+}
+
+const querySchema = z.object({
+  action: z.nativeEnum(DirectAction),
+  token: z.string(),
+  reason: z.string().optional(),
+  bookingUid: z.string(),
+  userId: z.string(),
```
Comment (zomars) [inline apps/web/pages/api/verify-booking-token.ts:23]: As tokens are unique then we can use them to identify the booking. We don't need the UID.

```suggestion
  action: z.nativeEnum(DirectAction),
  token: z.string(),
  reason: z.string().optional(),
  userId: z.string(),
```

File: apps/web/pages/api/verify-booking-token.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  const booking = await prisma.booking.findUniqueOrThrow({
+    where: { uid: bookingUid },
+  });
+
+  const secret = booking.bookingSecret as string;
+  const isValidToken = authenticator.check(token, secret);
+  if (!isValidToken) {
+    res.redirect(`/booking/${bookingUid}?error=${encodeURIComponent("Error confirming booking")}`);
+    return;
+  }
```
Comment (zomars) [inline apps/web/pages/api/verify-booking-token.ts:38]:
```suggestion
  const booking = await prisma.booking.findUnique({
    where: { oneTimePassword: token },
  });

  if (!booking) {
  	// Or a custom error page where we state that the booking doesn't exist or the token was invalid.
    res.redirect(`/booking/404`);
    return;
  }
```

**Review by zomars — CHANGES_REQUESTED** [review]
Thank you for your contribution @anikdhabal 

Leaving some feedback to address. 

I think we can simplify this even further.

I think naming the field `oneTimePassword` instead of `bookingSecret` is more intentional. Also `booking.bookingSecret` sounds redundant. We're are already in the booking model.

Using the authenticator library is overkill IMO. The token is random and secret enough and should be used as a disposable password anyways.

Just got the idea of testing then possibility of including a text field directly in the email to provide the rejection reason. Will test locally and keep you posted.

File: apps/web/pages/api/verify-booking-token.ts (hunk 1 above)
Comment (anikdhabal) [reply apps/web/pages/api/verify-booking-token.ts]: Yeah, it's not possible now. I missed it to remove

**Comment by anikdhabal** [comment]
Thanks @zomars for the review, yes we can simply it. Fixed it

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 4]
```diff
@@ -901,6 +902,7 @@ async function handler(
     platformRescheduleUrl,
     platformCancelUrl,
     platformBookingUrl,
+    bookingSecret: isConfirmedByDefault ? null : authenticator.generateSecret(),
```
[the remark this answers is not in the document]
Comment (anikdhabal) [reply packages/features/bookings/lib/handleNewBooking.ts]: But if the oneTimePassword value is null or undefined, then we have no password value to pass to the PI from the email template. props.calEvent.oneTimePassword is equal to undefined.

File: apps/web/pages/api/verify-booking-token.ts (hunk 2 above)
Comment (anikdhabal) [reply apps/web/pages/api/verify-booking-token.ts]: I think we need it. If the token is not valid, we don't get the booking UID, but the UID is needed to redirect to the /uid/404 page.

File: packages/emails/src/templates/OrganizerRequestEmail.tsx [hunk 5]
```diff
@@ -22,20 +22,22 @@ export const OrganizerRequestEmail = (props: React.ComponentProps<typeof Organiz
...
+          <CallToActionTable>
+            <CallToAction
+              label={props.calEvent.organizer.language.translate("confirm")}
+              href={`${actionHref}&action=accept`}
+              startIconName="confirmIcon"
+            />
+            <Separator />
+            <CallToAction
+              label={props.calEvent.organizer.language.translate("reject")}
+              // href={`${actionHref}&action=reject`}
```
Comment (zomars) [inline packages/emails/src/templates/OrganizerRequestEmail.tsx:35]: If omitted it will be a submit button for the form.

**Review by zomars — COMMENTED** [review]
Hey @anikdhabal made some changes. Can you verify if it's working properly?

**Comment by anikdhabal** [comment]
> Hey @anikdhabal made some changes. Can you verify if it's working properly?

Okay let me check

File: apps/web/pages/api/verify-booking-token.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  const { reason } = z.object({ reason: z.string().optional() }).parse(req.body);
+
+  const booking = await prisma.booking.findUnique({
+    where: { oneTimePassword: token },
+  });
+
+  if (!booking) {
+    res.redirect(`/booking/${bookingUid}?error=${encodeURIComponent("Error confirming booking")}`);
+    return;
+  }
```
Comment (anikdhabal) [inline apps/web/pages/api/verify-booking-token.ts:36]: @zomars why do we remove the 404 page?

Comment (zomars) [reply apps/web/pages/api/verify-booking-token.ts:36]: I think we can leave it like that for now.

Comment (zomars) [reply apps/web/pages/api/verify-booking-token.ts:36]: Also the 404 page didn't work as expected as it might conflict with UID

**Comment by anikdhabal** [comment]
https://github.com/user-attachments/assets/792f407e-afae-4cd2-913a-aeff7e26e1c7

@zomars Not working properly; I think it makes sense to keep the action href and redirect just as before.

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 7]
```diff
@@ -901,6 +901,7 @@ async function handler(
     platformRescheduleUrl,
     platformCancelUrl,
     platformBookingUrl,
+    oneTimePassword: isConfirmedByDefault ? null : undefined,
```
Comment (anikdhabal) [inline packages/features/bookings/lib/handleNewBooking.ts:904]: We need to update the one-time password after the booking, or we will get an undefined value in the email template. Currently, this is happening

File: packages/emails/src/components/BookingConfirmationForm.tsx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            textAlign: "left",
+            whiteSpace: "pre-wrap",
+            display: "block",
+          }}>
+          Reason for rejection &nbsp;
+          <small>(Optional)</small>
+        </label>
+        <textarea
+          placeholder="Why are you rejecting?"
+          style={{
```
Comment (anikdhabal) [inline packages/emails/src/components/BookingConfirmationForm.tsx:43]: Need to set this field name to 'reason.

**Comment by anikdhabal** [comment]
Screenshot 2024-08-23 123509
Screenshot 2024-08-23 123358

@zomars not intuitive, i think redirecting to our booking canceled page would be better, just like before. What do you think?

**Comment by anikdhabal** [comment]
@zomars take a look whenever got a chance

**Review by zomars — CHANGES_REQUESTED** [review]
Blocking to put a feature flag before merging #16038

File: packages/emails/src/templates/index.ts [hunk 9]
```diff
@@ -15,6 +15,7 @@ export { OrganizerCancelledEmail } from "./OrganizerCancelledEmail";
 export { OrganizerLocationChangeEmail } from "./OrganizerLocationChangeEmail";
 export { OrganizerPaymentRefundFailedEmail } from "./OrganizerPaymentRefundFailedEmail";
 export { OrganizerRequestEmail } from "./OrganizerRequestEmail";
+export { OrganizerRequestEmailV2 } from "./OrganizerRequestEmailV2";
```
Comment (zomars) [inline packages/emails/src/templates/index.ts:18]: Moved to a new template so we can enable it on demand @anikdhabal

**Comment by zomars** [comment]
Made new tests and it's working flawlessly. Now we can also enable this for specific users and teams so we can test on multiple email clients before enabling globally. cc @emrysal @keithwillcode 

https://github.com/user-attachments/assets/39dffe79-6efb-4641-9022-a97d93038092

**Review by zomars — APPROVED** [review]
Made new tests and it's working flawlessly. Now we can also enable this for specific users and teams so we can test on multiple email clients before enabling globally. cc @emrysal @keithwillcode 

https://github.com/user-attachments/assets/39dffe79-6efb-4641-9022-a97d93038092

File: packages/emails/src/templates/index.ts (hunk 9 above)
Comment (anikdhabal) [reply packages/emails/src/templates/index.ts:18]: Great workaround @zomars 🚀

**Pull request merged**